crowbar_framework/app/controllers/api/upgrade_controller.rb

```diff
@@ -62,6 +62,16 @@ def services
     end
   end
 
+  api :POST, "/api/upgrade/nodes", "Initiate the upgrade of all nodes"
+  api_version "2.0"
+  # This is gonna initiate the upgrade of all nodes.
+  # The method runs asynchronously, so there's a need to poll for the status and possible errors
+  def nodes
+    # FIXME: implement this as asynchronous call!
```

Comment:
Right - when it's changed to asynchronous, the background job/thread/whatever can simply persist the error (to somewhere which any later status calls can read from) and then abort.  This could be done by calling a helper, or by raising an exception which gets caught and then the handler persists the error and aborts.